_posts/2023-11-10-automated-test-generation-chess-puzzles-with-dafny.markdown

```diff
@@ -0,0 +1,288 @@
+---
+layout: post
+title:  "Automated Test Generation: Chess Puzzles with Dafny"
+date:   2023-11-10 11:00:00 -0500
+author: Sasha Fedchin
+---
+Given all the incredible examples that showcase how powerful Dafny is (did you know that part of Dafny is written in Dafny?), you might think that verification can fully replace testing. And in many cases it does! However if, like a number of industrial Dafny users, you want to integrate Dafny code with existing codebases, you will face challenges that verification alone might not solve:
+
+- Perhaps your code is actually a reference implementation for another program and you want to test the equivalence between the two.
+- You might be calling out to external libraries, which are not themselves verified, and would like to test them by calling your Dafny code.
+- You have written high-level specifications for your code and are about to embark on a proof but want to have some initial assurance that your yet-to-be-verified implementation does conform to these specifications. 
+
+In this blog post, I show how Dafny’s built-in automated test generation can be useful in these situations. In the example I use, Dafny has to position pieces on a chess board in such a way that certain constraints are satisfied. This is exactly what I want to illustrate: chess boards will be generated tests, and constraints will be the program we want to test. A brute force enumeration or fuzzing might not be up to the task here, since the number of ways in which you can arrange pieces on the board is astronomically high, even under the constraints I add to the problem. Dafny, however, can use the verifier to generate tests and is, therefore, much more efficient in finding solutions. 
```

Comment:
> In this blog post

Anywhere in this blog post? Consider being more specific: "In the following sections," 

Also, whether it's a blog post or a newspaper article doesn't seem relevant. 

> I show how

Consider replacing "I" with "we" where "we" it means yourself and the reader, "we will see how Dafny's built-in.. ". 

Looking at some of the other blog posts, like https://dafny.org/blog/2023/08/14/clear-specification-and-implementation/, they don't use I at all.

Also consider using the text itself as the subject. "The following sections explain how Dafny’s built-in automated test generation can help with these scenarios"

Reply:
Have changed "In this blog, I..." to "The following sections explain". 

I removed a few uses of first person in the latest set of changes, but I think even amongst the existing blogpost, this is more a matter of preference, e.g. [making verification compelling blogpost](https://dafny.org/blog/2023/04/19/making-verification-compelling-visual-verification-feedback-for-dafny/#step-0-0) uses first person quite often.